### DISCUSSION TITLE
What is the purpose of PrimordialHandler class and while default indent level is set to zero?

### DISCUSSION THREAD

**Discussion by kalpadiptyaroy** [checkstyle/checkstyle#14498] [Q&A] [answered]
I am raising this question due to https://github.com/checkstyle/checkstyle/pull/14149#issuecomment-1946129857 

consider the code below.

https://github.com/checkstyle/checkstyle/blob/0723bc17b4e679fbb430c516181dd29ca91ff42a/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/PrimordialHandler.java#L47-L50

I wanted to enquire - what is purpose of hardcoding `0` as the default indent level?
And also, I want to understand the purpose of this class `PrimordialHandler` ?

**Comment by romani** [comment]
Try to go to git history to get links to issues and PRs when it was introduced.

**Comment by nrmancuso** [comment, marked as answer]
I would assume that the name `PrimordialHandler` is to hint that this is the base/initial/fundamental indentation handler; therefore we would always want the base indentation to be zero.

**Reply by rnveach** [reply]
The class literally says it is a no-op class. This means no indentation increase for whatever it points out.
https://github.com/checkstyle/checkstyle/blob/a4411aa08c3a9744139067dc38ced282adbf9fcc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/PrimordialHandler.java#L23

**Comment by kalpadiptyaroy** [comment]
Fine. I understood the purpose.
Now, I would like to ask whether the following behavior is the desired behavior or is it not?

When indentation check is triggered - then the program flows in the following way --

1. `checkModifiers` method is called.

https://github.com/checkstyle/checkstyle/blob/a4411aa08c3a9744139067dc38ced282adbf9fcc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AbstractExpressionHandler.java#L496-L508

2. `getIndent` method is called.

https://github.com/checkstyle/checkstyle/blob/a4411aa08c3a9744139067dc38ced282adbf9fcc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AbstractExpressionHandler.java#L83-L88

3. `getIndentImpl` method is called.

https://github.com/checkstyle/checkstyle/blob/a4411aa08c3a9744139067dc38ced282adbf9fcc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AbstractExpressionHandler.java#L95-L97

Now, here this `return parent.getSuggestedChildIndent(this);` is has an overridden implementation and it executes the following method from `PrimordialHandler` for all the handlers that are executed by `TreeWalker`.

https://github.com/checkstyle/checkstyle/blob/a4411aa08c3a9744139067dc38ced282adbf9fcc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/PrimordialHandler.java#L48-L50

Now, the question is: should it be like this: 
that the IndentLevel's value is initially set to zero obeying the implementation of primordial handler, but when we are using another handler (e.g. ClassDefHandler .... etc) then, the indentlevel parameters' value should be overridden as per the configuration of that handler?

    --------------------- UPDATE --------------------->>

I have further investigated and found the `PrimordialHandler` is injected as the parent of each handlers, here in IndentationCheck.

https://github.com/checkstyle/checkstyle/blob/a4411aa08c3a9744139067dc38ced282adbf9fcc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheck.java#L426-L430

**Reply by nrmancuso** [reply]
> behavior

Figuring out how the code here is supposed to behave is a fool's errand, IMO. This is our most complicated (and buggy) check. It is better to focus on actual behavior, what we need to do to get to the expected check behavior, and thorough regression testing.

**Reply by kalpadiptyaroy** [reply]
may be you are right - but I have made certain fixes - explore here https://github.com/checkstyle/checkstyle/pull/14149#issuecomment-1971604698
